File: query-primitives/src/main/java/io/shiftleft/queryprimitives/CpgLoader.java
package io.shiftleft.queryprimitives;

import io.shiftleft.queryprimitives.cpgloading.proto.ProtoCpgLoader;
import io.shiftleft.queryprimitives.cpgloading.tinker.TinkerCpgLoader;
import io.shiftleft.queryprimitives.cpgloading.tinker.TinkerFormat;
import io.shiftleft.queryprimitives.steps.starters.Cpg;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;
import org.apache.tinkerpop.gremlin.structure.Vertex;
import org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph;

public class CpgLoader {

  /**
   * Load a Code Property Graph
   * @param filename name of file that stores the code property graph
   * @param format input format of the CPG
   */
  public static Cpg loadCodePropertyGraph(String filename, CpgInputFormat format) {
    Cpg cpg;

    switch (format) {
      case PROTOZIP:
        cpg = ProtoCpgLoader.getInstance().loadFromProtoZip(filename);
        break;
      case GRYO:
        cpg = TinkerCpgLoader.loadFromTinkerFormat(filename, TinkerFormat.GRYO);
        break;
      case TINKERPOP:
        cpg = TinkerCpgLoader.loadFromTinkerFormat(filename, TinkerFormat.GRAPHML);
        break;
      case GRAPHML:
        cpg = TinkerCpgLoader.loadFromTinkerFormat(filename, TinkerFormat.GRAPHML);
        break;
      default:
        // Match is exhaustive, but javac still complains
        cpg = null;
        break;
    }

    return cpg;
  }

}

File: query-primitives/src/main/scala/io/shiftleft/queryprimitives/CpgLoader.scala
package io.shiftleft.queryprimitives

import io.shiftleft.codepropertygraph.generated.NodeKeys
import io.shiftleft.cpgloading.ProtoCpgLoader
import io.shiftleft.cpgloading.OnDiskOverflowConfig
import io.shiftleft.queryprimitives.steps.starters.Cpg
import org.apache.logging.log4j.LogManager
import org.apache.tinkerpop.gremlin.structure.Vertex
import org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph
import scala.compat.java8.OptionConverters._

object CpgLoader {
  private val logger = LogManager.getLogger(classOf[CpgLoader])

  /**
    * Load a Code Property Graph
    *
    * @param filename      name of file that stores the code property graph
    * @param createIndexes create indexes after loading the file
    */
  def loadCodePropertyGraph(filename: String,
                            createIndices: Boolean = true,
                            onDiskOverflowConfig: Option[OnDiskOverflowConfig] = None): Cpg = {
    var cpg: Cpg = null
    logger.debug("Loading " + filename)
    cpg = ProtoCpgLoader.loadFromProtoZip(filename, onDiskOverflowConfig.asJava)
    if (createIndices) { createIndexes(cpg) }
    cpg
  }

  /**
    * Create any indexes necessary for quick access.
    *
    * @param cpg the CPG to create indexes in
    */
  def createIndexes(cpg: Cpg): Unit =
    cpg.graph.asInstanceOf[TinkerGraph].createIndex(NodeKeys.FULL_NAME.name, classOf[Vertex])

  def addOverlays(overlayFilenames: Seq[String], cpg: Cpg) =
    overlayFilenames.foreach { overlayFilename =>
      CpgOverlayLoader.load(overlayFilename, cpg)
    }

}

class CpgLoader {}
